TEXT CLEANING

In [32]:
import pandas as pd

In [33]:
df = pd.read_csv("../data/train.csv")
df.head()

,review,sentiment
0,I caught this little gem totally by accident b...,positive
1,I can't believe that I let myself into this mo...,negative
2,*spoiler alert!* it just gets to me the nerve ...,negative
3,If there's one thing I've learnt from watching...,negative
4,"I remember when this was in theaters, reviews ...",negative


In [34]:
import re

def clean_text(text):
    text = text.lower()
    text = re.sub(r"[^a-z0-9\s]", "", text)
    return text
df["clean_review"] = df["review"].apply(clean_text)
df["clean_review"].iloc[0]

'i caught this little gem totally by accident back in 1980 or 81 i was at a revival theatre to see two old silly scifi movies the theatre was packed full and with no warning they showed a bunch of scifi short spoofs to get us in the mood most were somewhat amusing but this came on and within seconds the audience was in hysterics the biggest laugh came when they showed princess laia having huge cinnamon buns instead of hair on her head she looks at the camera gives a grim smile and nods that made it even funnier you gotta see chewabacca played by what looks like a muppet it was extremely silly and stupidbut i couldnt stop laughing most of the dialogue was drowned out because of all the laughter also if you know star wars pretty well its even funnierthey deliberately poke fun at some of the dialogue this really works with an audience a definite 10'

TOKENIZATION

In [37]:
sample_text = df["clean_review"].iloc[0]
sample_text
tokens = sample_text.split()
tokens

['i',
 'caught',
 'this',
 'little',
 'gem',
 'totally',
 'by',
 'accident',
 'back',
 'in',
 '1980',
 'or',
 '81',
 'i',
 'was',
 'at',
 'a',
 'revival',
 'theatre',
 'to',
 'see',
 'two',
 'old',
 'silly',
 'scifi',
 'movies',
 'the',
 'theatre',
 'was',
 'packed',
 'full',
 'and',
 'with',
 'no',
 'warning',
 'they',
 'showed',
 'a',
 'bunch',
 'of',
 'scifi',
 'short',
 'spoofs',
 'to',
 'get',
 'us',
 'in',
 'the',
 'mood',
 'most',
 'were',
 'somewhat',
 'amusing',
 'but',
 'this',
 'came',
 'on',
 'and',
 'within',
 'seconds',
 'the',
 'audience',
 'was',
 'in',
 'hysterics',
 'the',
 'biggest',
 'laugh',
 'came',
 'when',
 'they',
 'showed',
 'princess',
 'laia',
 'having',
 'huge',
 'cinnamon',
 'buns',
 'instead',
 'of',
 'hair',
 'on',
 'her',
 'head',
 'she',
 'looks',
 'at',
 'the',
 'camera',
 'gives',
 'a',
 'grim',
 'smile',
 'and',
 'nods',
 'that',
 'made',
 'it',
 'even',
 'funnier',
 'you',
 'gotta',
 'see',
 'chewabacca',
 'played',
 'by',
 'what',
 'looks',
 'like

In [36]:
df["tokens"] = df["clean_review"].apply(lambda x: x.lower().split())
df[["clean_review", "tokens"]].head()

,clean_review,tokens
0,i caught this little gem totally by accident b...,"[i, caught, this, little, gem, totally, by, ac..."
1,i cant believe that i let myself into this mov...,"[i, cant, believe, that, i, let, myself, into,..."
2,spoiler alert it just gets to me the nerve som...,"[spoiler, alert, it, just, gets, to, me, the, ..."
3,if theres one thing ive learnt from watching g...,"[if, theres, one, thing, ive, learnt, from, wa..."
4,i remember when this was in theaters reviews s...,"[i, remember, when, this, was, in, theaters, r..."


In [6]:
type(df["tokens"].iloc[0])

list

The text was converted to lowercase and tokenized by splitting each review into individual words.

STOPWORDS REMOVAL

In [38]:
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

In [39]:
list(ENGLISH_STOP_WORDS)[:20]

['ourselves',
 'latter',
 'may',
 'us',
 'those',
 'however',
 'this',
 'being',
 'few',
 'if',
 'but',
 'no',
 'whole',
 'many',
 'yourselves',
 'be',
 'cant',
 'find',
 'system',
 'or']

In [40]:
sample_tokens = df["tokens"].iloc[0]
sample_tokens

['i',
 'caught',
 'this',
 'little',
 'gem',
 'totally',
 'by',
 'accident',
 'back',
 'in',
 '1980',
 'or',
 '81',
 'i',
 'was',
 'at',
 'a',
 'revival',
 'theatre',
 'to',
 'see',
 'two',
 'old',
 'silly',
 'scifi',
 'movies',
 'the',
 'theatre',
 'was',
 'packed',
 'full',
 'and',
 'with',
 'no',
 'warning',
 'they',
 'showed',
 'a',
 'bunch',
 'of',
 'scifi',
 'short',
 'spoofs',
 'to',
 'get',
 'us',
 'in',
 'the',
 'mood',
 'most',
 'were',
 'somewhat',
 'amusing',
 'but',
 'this',
 'came',
 'on',
 'and',
 'within',
 'seconds',
 'the',
 'audience',
 'was',
 'in',
 'hysterics',
 'the',
 'biggest',
 'laugh',
 'came',
 'when',
 'they',
 'showed',
 'princess',
 'laia',
 'having',
 'huge',
 'cinnamon',
 'buns',
 'instead',
 'of',
 'hair',
 'on',
 'her',
 'head',
 'she',
 'looks',
 'at',
 'the',
 'camera',
 'gives',
 'a',
 'grim',
 'smile',
 'and',
 'nods',
 'that',
 'made',
 'it',
 'even',
 'funnier',
 'you',
 'gotta',
 'see',
 'chewabacca',
 'played',
 'by',
 'what',
 'looks',
 'like

In [41]:
filtered_tokens = [w for w in sample_tokens if w not in ENGLISH_STOP_WORDS]
filtered_tokens

['caught',
 'little',
 'gem',
 'totally',
 'accident',
 '1980',
 '81',
 'revival',
 'theatre',
 'old',
 'silly',
 'scifi',
 'movies',
 'theatre',
 'packed',
 'warning',
 'showed',
 'bunch',
 'scifi',
 'short',
 'spoofs',
 'mood',
 'somewhat',
 'amusing',
 'came',
 'seconds',
 'audience',
 'hysterics',
 'biggest',
 'laugh',
 'came',
 'showed',
 'princess',
 'laia',
 'having',
 'huge',
 'cinnamon',
 'buns',
 'instead',
 'hair',
 'head',
 'looks',
 'camera',
 'gives',
 'grim',
 'smile',
 'nods',
 'funnier',
 'gotta',
 'chewabacca',
 'played',
 'looks',
 'like',
 'muppet',
 'extremely',
 'silly',
 'stupidbut',
 'stop',
 'laughing',
 'dialogue',
 'drowned',
 'laughter',
 'know',
 'star',
 'wars',
 'pretty',
 'funnierthey',
 'deliberately',
 'poke',
 'fun',
 'dialogue',
 'really',
 'works',
 'audience',
 'definite',
 '10']

In [42]:
df["tokens_no_stopwords"] = df["tokens"].apply(
    lambda tokens: [w for w in tokens if w not in ENGLISH_STOP_WORDS]
)

df[["tokens", "tokens_no_stopwords"]].head()

,tokens,tokens_no_stopwords
0,"[i, caught, this, little, gem, totally, by, ac...","[caught, little, gem, totally, accident, 1980,..."
1,"[i, cant, believe, that, i, let, myself, into,...","[believe, let, movie, accomplish, favor, frien..."
2,"[spoiler, alert, it, just, gets, to, me, the, ...","[spoiler, alert, just, gets, nerve, people, re..."
3,"[if, theres, one, thing, ive, learnt, from, wa...","[theres, thing, ive, learnt, watching, george,..."
4,"[i, remember, when, this, was, in, theaters, r...","[remember, theaters, reviews, said, horrible, ..."


Common English stopwords were removed in order to reduce noise and have only informative words that contribute to classification

STEMMING

In [43]:
from nltk.stem import PorterStemmer

In [44]:
stemmer = PorterStemmer()

In [45]:
sample_tokens = df["tokens_no_stopwords"].iloc[0]
sample_tokens


['caught',
 'little',
 'gem',
 'totally',
 'accident',
 '1980',
 '81',
 'revival',
 'theatre',
 'old',
 'silly',
 'scifi',
 'movies',
 'theatre',
 'packed',
 'warning',
 'showed',
 'bunch',
 'scifi',
 'short',
 'spoofs',
 'mood',
 'somewhat',
 'amusing',
 'came',
 'seconds',
 'audience',
 'hysterics',
 'biggest',
 'laugh',
 'came',
 'showed',
 'princess',
 'laia',
 'having',
 'huge',
 'cinnamon',
 'buns',
 'instead',
 'hair',
 'head',
 'looks',
 'camera',
 'gives',
 'grim',
 'smile',
 'nods',
 'funnier',
 'gotta',
 'chewabacca',
 'played',
 'looks',
 'like',
 'muppet',
 'extremely',
 'silly',
 'stupidbut',
 'stop',
 'laughing',
 'dialogue',
 'drowned',
 'laughter',
 'know',
 'star',
 'wars',
 'pretty',
 'funnierthey',
 'deliberately',
 'poke',
 'fun',
 'dialogue',
 'really',
 'works',
 'audience',
 'definite',
 '10']

In [46]:
stemmed_tokens = [stemmer.stem(w) for w in sample_tokens]
stemmed_tokens

['caught',
 'littl',
 'gem',
 'total',
 'accid',
 '1980',
 '81',
 'reviv',
 'theatr',
 'old',
 'silli',
 'scifi',
 'movi',
 'theatr',
 'pack',
 'warn',
 'show',
 'bunch',
 'scifi',
 'short',
 'spoof',
 'mood',
 'somewhat',
 'amus',
 'came',
 'second',
 'audienc',
 'hyster',
 'biggest',
 'laugh',
 'came',
 'show',
 'princess',
 'laia',
 'have',
 'huge',
 'cinnamon',
 'bun',
 'instead',
 'hair',
 'head',
 'look',
 'camera',
 'give',
 'grim',
 'smile',
 'nod',
 'funnier',
 'gotta',
 'chewabacca',
 'play',
 'look',
 'like',
 'muppet',
 'extrem',
 'silli',
 'stupidbut',
 'stop',
 'laugh',
 'dialogu',
 'drown',
 'laughter',
 'know',
 'star',
 'war',
 'pretti',
 'funnierthey',
 'deliber',
 'poke',
 'fun',
 'dialogu',
 'realli',
 'work',
 'audienc',
 'definit',
 '10']

Stemming significantly reduces words to their root forms, often producing non-dictionary tokens (e.g., theatre → theatr, little → littl).
While this approach effectively normalizes word variants, it may remove semantic information.

In [47]:
df["tokens_stemmed"] = df["tokens_no_stopwords"].apply(
    lambda tokens: [stemmer.stem(w) for w in tokens]
)

df[["tokens_no_stopwords", "tokens_stemmed"]].head()

,tokens_no_stopwords,tokens_stemmed
0,"[caught, little, gem, totally, accident, 1980,...","[caught, littl, gem, total, accid, 1980, 81, r..."
1,"[believe, let, movie, accomplish, favor, frien...","[believ, let, movi, accomplish, favor, friend,..."
2,"[spoiler, alert, just, gets, nerve, people, re...","[spoiler, alert, just, get, nerv, peopl, remak..."
3,"[theres, thing, ive, learnt, watching, george,...","[there, thing, ive, learnt, watch, georg, rome..."
4,"[remember, theaters, reviews, said, horrible, ...","[rememb, theater, review, said, horribl, didnt..."


In [48]:
df["tokens_stemmed"].iloc[0]

['caught',
 'littl',
 'gem',
 'total',
 'accid',
 '1980',
 '81',
 'reviv',
 'theatr',
 'old',
 'silli',
 'scifi',
 'movi',
 'theatr',
 'pack',
 'warn',
 'show',
 'bunch',
 'scifi',
 'short',
 'spoof',
 'mood',
 'somewhat',
 'amus',
 'came',
 'second',
 'audienc',
 'hyster',
 'biggest',
 'laugh',
 'came',
 'show',
 'princess',
 'laia',
 'have',
 'huge',
 'cinnamon',
 'bun',
 'instead',
 'hair',
 'head',
 'look',
 'camera',
 'give',
 'grim',
 'smile',
 'nod',
 'funnier',
 'gotta',
 'chewabacca',
 'play',
 'look',
 'like',
 'muppet',
 'extrem',
 'silli',
 'stupidbut',
 'stop',
 'laugh',
 'dialogu',
 'drown',
 'laughter',
 'know',
 'star',
 'war',
 'pretti',
 'funnierthey',
 'deliber',
 'poke',
 'fun',
 'dialogu',
 'realli',
 'work',
 'audienc',
 'definit',
 '10']

LEMMATIZATION

In [49]:
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Visnja\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [50]:
from nltk.stem import WordNetLemmatizer

In [51]:
lemmatizer = WordNetLemmatizer()

In [52]:
sample_tokens = df["tokens_no_stopwords"].iloc[0]
sample_tokens

['caught',
 'little',
 'gem',
 'totally',
 'accident',
 '1980',
 '81',
 'revival',
 'theatre',
 'old',
 'silly',
 'scifi',
 'movies',
 'theatre',
 'packed',
 'warning',
 'showed',
 'bunch',
 'scifi',
 'short',
 'spoofs',
 'mood',
 'somewhat',
 'amusing',
 'came',
 'seconds',
 'audience',
 'hysterics',
 'biggest',
 'laugh',
 'came',
 'showed',
 'princess',
 'laia',
 'having',
 'huge',
 'cinnamon',
 'buns',
 'instead',
 'hair',
 'head',
 'looks',
 'camera',
 'gives',
 'grim',
 'smile',
 'nods',
 'funnier',
 'gotta',
 'chewabacca',
 'played',
 'looks',
 'like',
 'muppet',
 'extremely',
 'silly',
 'stupidbut',
 'stop',
 'laughing',
 'dialogue',
 'drowned',
 'laughter',
 'know',
 'star',
 'wars',
 'pretty',
 'funnierthey',
 'deliberately',
 'poke',
 'fun',
 'dialogue',
 'really',
 'works',
 'audience',
 'definite',
 '10']

In [53]:
lemmatized_tokens = [lemmatizer.lemmatize(w) for w in sample_tokens]
lemmatized_tokens

['caught',
 'little',
 'gem',
 'totally',
 'accident',
 '1980',
 '81',
 'revival',
 'theatre',
 'old',
 'silly',
 'scifi',
 'movie',
 'theatre',
 'packed',
 'warning',
 'showed',
 'bunch',
 'scifi',
 'short',
 'spoof',
 'mood',
 'somewhat',
 'amusing',
 'came',
 'second',
 'audience',
 'hysteric',
 'biggest',
 'laugh',
 'came',
 'showed',
 'princess',
 'laia',
 'having',
 'huge',
 'cinnamon',
 'bun',
 'instead',
 'hair',
 'head',
 'look',
 'camera',
 'give',
 'grim',
 'smile',
 'nod',
 'funnier',
 'gotta',
 'chewabacca',
 'played',
 'look',
 'like',
 'muppet',
 'extremely',
 'silly',
 'stupidbut',
 'stop',
 'laughing',
 'dialogue',
 'drowned',
 'laughter',
 'know',
 'star',
 'war',
 'pretty',
 'funnierthey',
 'deliberately',
 'poke',
 'fun',
 'dialogue',
 'really',
 'work',
 'audience',
 'definite',
 '10']

In [55]:
df["tokens_lemmatized"] = df["tokens_no_stopwords"].apply(
    lambda tokens: [lemmatizer.lemmatize(w) for w in tokens]
)

df[["tokens_no_stopwords", "tokens_lemmatized"]].head()

,tokens_no_stopwords,tokens_lemmatized
0,"[caught, little, gem, totally, accident, 1980,...","[caught, little, gem, totally, accident, 1980,..."
1,"[believe, let, movie, accomplish, favor, frien...","[believe, let, movie, accomplish, favor, frien..."
2,"[spoiler, alert, just, gets, nerve, people, re...","[spoiler, alert, just, get, nerve, people, rem..."
3,"[theres, thing, ive, learnt, watching, george,...","[there, thing, ive, learnt, watching, george, ..."
4,"[remember, theaters, reviews, said, horrible, ...","[remember, theater, review, said, horrible, di..."


In [56]:
sample_tokens = df["tokens_no_stopwords"].iloc[4]
sample_stems = df["tokens_stemmed"].iloc[4]
sample_lemmas = df["tokens_lemmatized"].iloc[4]

comparison_df = pd.DataFrame({
    "Original": sample_tokens,
    "Stemmed": sample_stems,
    "Lemmatized": sample_lemmas
})

comparison_df

,Original,Stemmed,Lemmatized
0,remember,rememb,remember
1,theaters,theater,theater
2,reviews,review,review
3,said,said,said
4,horrible,horribl,horrible
...,...,...,...
78,thought,thought,thought
79,worth,worth,worth
80,checking,check,checking
81,catch,catch,catch


Stemming aggressively modiefies words to their root forms (e.g., theaters → theatr).
In contrast, lemmatization maps words to their proper dictionary forms (theaters → theater), preserving semantic meaning.

Stemming vs Lemmatization:
While stemming provides aggressive normalization and significantly reduces vocabulary size(e.g., theaters → theatr), it often produces non-interpretable tokens and may lose semantic information.
Lemmatization keeps valid word forms and semantic meaning, resulting in cleaner text representations.


VECTORIZATION

In [58]:
from sklearn.feature_extraction.text import CountVectorizer

In [57]:
df["text_for_vectorizer"] = df["tokens_lemmatized"].apply(lambda x: " ".join(x))
df["text_for_vectorizer"].head()

0    caught little gem totally accident 1980 81 rev...
1    believe let movie accomplish favor friend ask ...
2    spoiler alert just get nerve people remake use...
3    there thing ive learnt watching george romeros...
4    remember theater review said horrible didnt th...
Name: text_for_vectorizer, dtype: object

In [59]:
bow_vectorizer = CountVectorizer(max_features=5000)

In [ ]:
X_bow = bow_vectorizer.fit_transform(df["text_for_vectorizer"])
X_bow.shape #40.000 reviews and 5.000 words in the vocabulary

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 2780038 stored elements and shape (40000, 5000)>

Bag of Words represents each review as a vector of word occurrence counts, where each feature corresponds to a word from the vocabulary

In [74]:
df["sentiment"].unique()

array(['positive', 'negative'], dtype=object)

In [75]:
y = df["sentiment"].map({"positive": 1, "negative": 0})

In [62]:
X_bow.shape, y.shape

((40000, 5000), (40000,))

TF-IDF

In [64]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [65]:
tfidf_vectorizer = TfidfVectorizer(max_features=5000)

In [68]:
X_tfidf = tfidf_vectorizer.fit_transform(df["text_for_vectorizer"])
X_tfidf.shape

(40000, 5000)

TF-IDF weighs words by considering both their frequency and their rarity,reducing the impact of common but less informative words.

BoW vs TF-IDF:
Bag of Words only counts how many times a word appears in a review.
TF-IDF reduces the importance of very common words and gives more weight to words that are more useful for classifying sentiment.
For this reason, TF-IDF is better option sentiment classification tasks.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_uni = TfidfVectorizer(
    max_features=5000,
    ngram_range=(1, 1)
)

X_tfidf_uni = tfidf_uni.fit_transform(df["text_for_vectorizer"])

In [77]:
tfidf_bi = TfidfVectorizer(
    max_features=5000,
    ngram_range=(1, 2)
)

X_tfidf_bi = tfidf_bi.fit_transform(df["text_for_vectorizer"])

In [71]:
tfidf_bi.get_feature_names_out()[:20]

array(['10', '10 10', '10 minute', '10 year', '100', '1010', '11', '110',
       '12', '13', '14', '15', '15 minute', '15 year', '16', '17', '18',
       '1930s', '1940s', '1950s'], dtype=object)

In [78]:
y.isna().sum()

np.int64(0)

In [79]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

X_train, X_val, y_train, y_val = train_test_split(
    X_tfidf_uni, y, test_size=0.2, random_state=42
)

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)
preds = model.predict(X_val)

acc_uni = accuracy_score(y_val, preds)
acc_uni


0.880625

In [80]:
X_train, X_val, y_train, y_val = train_test_split(
    X_tfidf_bi, y, test_size=0.2, random_state=42
)

model.fit(X_train, y_train)
preds = model.predict(X_val)

acc_bi = accuracy_score(y_val, preds)
acc_bi

0.8825

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

X_train, X_val, y_train, y_val = train_test_split(
    X_tfidf_uni, y, test_size=0.2, random_state=42
)

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)
preds = model.predict(X_val)

acc_uni = accuracy_score(y_val, preds)
acc_uni


0.880625

I compared TF-IDF using unigrams with TF-IDF using unigrams and bigrams.
Including bigrams allowed the model to capture short word combinations such as “not good” or “very bad”, which can carry sentiment information.
Although the performance improvement was modest (accuracy increased from 0.881 to 0.883), the bigram-based TF-IDF representation was selected for the final model as it provided a richer textual representation without a significant increase in complexity.

The preprocessing decisions implemented in this notebook are applied in the training pipeline.